SavepointMockServerTest is flaky

```
Error:  Tests run: 72, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.170 s <<< FAILURE! -- in com.google.cloud.spanner.connection.SavepointMockServerTest
Error:  com.google.cloud.spanner.connection.SavepointMockServerTest.testKeepAlive[dialect = POSTGRESQL, useVirtualThreads = true] -- Time elapsed: 0.037 s <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<2>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at com.google.cloud.spanner.connection.SavepointMockServerTest.testKeepAlive(SavepointMockServerTest.java:722)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:49)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:120)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:95)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:69)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
```

--- a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ReadWriteTransaction.java
+++ b/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ReadWriteTransaction.java
@@ -461,7 +461,10 @@ class ReadWriteTransaction extends AbstractMultiUseTransaction {
                 CallType.SYNC,
                 SELECT1_STATEMENT,
                 AnalyzeMode.NONE,
-                Options.tag("connection.transaction-keep-alive"));
+                Options.tag(
+                    System.getProperty(
+                        "spanner.connection.keep_alive_query_tag",
+                        "connection.transaction-keep-alive")));
         future.addListener(
             ReadWriteTransaction.this::maybeScheduleKeepAlivePing, MoreExecutors.directExecutor());
       }

--- a/google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/SavepointMockServerTest.java
+++ b/google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/SavepointMockServerTest.java
@@ -682,14 +682,16 @@ public class SavepointMockServerTest extends AbstractMockServerTest {
 
   @Test
   public void testKeepAlive() throws InterruptedException, TimeoutException {
+    String keepAliveTag = "test_keep_alive_tag";
     System.setProperty("spanner.connection.keep_alive_interval_millis", "1");
+    System.setProperty("spanner.connection.keep_alive_query_tag", keepAliveTag);
     try (Connection connection = createConnection()) {
       connection.setSavepointSupport(SavepointSupport.ENABLED);
       connection.setKeepTransactionAlive(true);
       // Start a transaction by executing a statement.
       connection.execute(INSERT_STATEMENT);
       // Verify that we get a keep-alive request.
-      verifyHasKeepAliveRequest();
+      verifyHasKeepAliveRequest(keepAliveTag);
       // Set a savepoint, execute another statement, and rollback to the savepoint.
       // The keep-alive should not be sent after the transaction has been rolled back to the
       // savepoint.
@@ -697,37 +699,22 @@ public class SavepointMockServerTest extends AbstractMockServerTest {
       connection.execute(INSERT_STATEMENT);
       connection.rollbackToSavepoint("s1");
       mockSpanner.waitForRequestsToContain(RollbackRequest.class, 1000L);
-      // Wait until all keep-alive requests that were in flight have finished.
-      try {
-        mockSpanner.waitForRequestsToContain(
-            r -> {
-              if (!(r instanceof ExecuteSqlRequest)) {
-                return false;
-              }
-              ExecuteSqlRequest request = (ExecuteSqlRequest) r;
-              return request.getSql().equals("SELECT 1")
-                  && request
-                      .getRequestOptions()
-                      .getRequestTag()
-                      .equals("connection.transaction-keep-alive");
-            },
-            200L);
-      } catch (TimeoutException ignore) {
-      }
+      String keepAliveTagAfterRollback = "test_keep_alive_tag_after_rollback";
+      System.setProperty("spanner.connection.keep_alive_query_tag", keepAliveTagAfterRollback);
 
-      // Verify that we don't get any keep-alive requests from this point.
-      mockSpanner.clearRequests();
+      // Verify that we don't get any new keep-alive requests from this point.
       Thread.sleep(2L);
-      assertEquals(0, countKeepAliveRequest());
+      assertEquals(0, countKeepAliveRequest(keepAliveTagAfterRollback));
       // Resume the transaction and verify that we get a keep-alive again.
       connection.execute(INSERT_STATEMENT);
-      verifyHasKeepAliveRequest();
+      verifyHasKeepAliveRequest(keepAliveTagAfterRollback);
     } finally {
       System.clearProperty("spanner.connection.keep_alive_interval_millis");
+      System.clearProperty("spanner.connection.keep_alive_query_tag");
     }
   }
 
-  private void verifyHasKeepAliveRequest() throws InterruptedException, TimeoutException {
+  private void verifyHasKeepAliveRequest(String tag) throws InterruptedException, TimeoutException {
     mockSpanner.waitForRequestsToContain(
         r -> {
           if (!(r instanceof ExecuteSqlRequest)) {
@@ -735,23 +722,17 @@ public class SavepointMockServerTest extends AbstractMockServerTest {
           }
           ExecuteSqlRequest request = (ExecuteSqlRequest) r;
           return request.getSql().equals("SELECT 1")
-              && request
-                  .getRequestOptions()
-                  .getRequestTag()
-                  .equals("connection.transaction-keep-alive");
+              && request.getRequestOptions().getRequestTag().equals(tag);
         },
         1000L);
   }
 
-  private long countKeepAliveRequest() {
+  private long countKeepAliveRequest(String tag) {
     return mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).stream()
         .filter(
             request ->
                 request.getSql().equals("SELECT 1")
-                    && request
-                        .getRequestOptions()
-                        .getRequestTag()
-                        .equals("connection.transaction-keep-alive"))
+                    && request.getRequestOptions().getRequestTag().equals(tag))
         .count();
   }
 }

--- a/google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/SavepointMockServerTest.java
+++ b/google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/SavepointMockServerTest.java
@@ -697,8 +697,7 @@ public class SavepointMockServerTest extends AbstractMockServerTest {
       connection.execute(INSERT_STATEMENT);
       connection.rollbackToSavepoint("s1");
       mockSpanner.waitForRequestsToContain(RollbackRequest.class, 1000L);
-      // Wait for up to 2 milliseconds to make sure that any keep-alive requests that were in flight
-      // have finished.
+      // Wait until all keep-alive requests that were in flight have finished.
       try {
         mockSpanner.waitForRequestsToContain(
             r -> {
@@ -712,7 +711,7 @@ public class SavepointMockServerTest extends AbstractMockServerTest {
                       .getRequestTag()
                       .equals("connection.transaction-keep-alive");
             },
-            2L);
+            200L);
       } catch (TimeoutException ignore) {
       }
 

